### PR TITLE
#163426480 Implement user registration with email verification

### PR DESCRIPTION
**Pull request by ozone4real**
#### What does this PR do?
It implements user registration with email verification
#### Description of Task to be completed?
- Design sign up form.
- Display descriptive validation error messages when a user enters unacceptable data.
- Display error response messages to the user  on unsuccessful registration
- Login the user on successful registration
- Implement email verification for signed up users.
- Implement user authentication.

Screen Shot 2019-03-18 at 2 13 58 PM

Screen Shot 2019-03-18 at 2 13 42 PM

Screen Shot 2019-03-18 at 2 11 35 PM

#### What are the relevant pivotal tracker stories?
[163426480](https://www.pivotaltracker.com/story/show/163426480)

### CODE REVIEW

File: __test__/actions/authActions.test.js [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+    const expectedActions = [
+      { type: VERIFY_USER_REQUEST },
+      { type: VERIFY_USER_SUCCESS }
+    ];
+
+    await store.dispatch(verifyUser(token, { history }));
+    expect(store.getActions()).toEqual(expectedActions);
+  });
+
+  it('shoild create the SIGN_OUT_USER action when the user signs out', () => {
```
Comment (omenkish) [inline __test__/actions/authActions.test.js]: I believe this should be 'SHOULD' not 'shoild'

**Review by omenkish — DISMISSED** [review]

**Review by EfeAgare — DISMISSED** [review]

**Review by biolabalo — DISMISSED** [review]

**Review by divinediscipline — DISMISSED** [review]
Looks good to me

File: __test__/components/__snapshots__/indexPage.test.js.snap [hunk 2]
```diff
@@ -3472,7 +3473,8 @@ Array [
             Get Started for free today.
           </p>
           <button
-            className="sc-bwzfXH catjVP"
+            className="sc-htpNat bimmNg"
```
Comment (p-funky) [inline __test__/components/__snapshots__/indexPage.test.js.snap:3476]: What's up with these class names? Are they going to change all the time?

File: package.json [hunk 3]
```diff
@@ -55,10 +57,11 @@
     "react-redux": "^6.0.1",
     "react-router": "^4.3.1",
     "react-router-dom": "^4.3.1",
-    "react-svg-inline": "^2.1.1",
     "react-spinners": "^0.5.3",
+    "react-svg-inline": "^2.1.1",
     "react-tag-input": "^6.3.0",
     "react-test-renderer": "^16.8.4",
+    "react-toastify": "^4.5.2",
```
Comment (p-funky) [inline package.json:65]: I believe we already have a package we are using for toasts so are we preferring this to the other? If yes, then that one should be removed.

File: src/App.js [hunk 4]
```diff
@@ -1,24 +1,40 @@
+/* eslint-disable import/no-named-as-default */
```
Comment (p-funky) [inline src/App.js]: We are talking about removing the existing disables so we can't keep adding more without sufficient justification.

File: src/assets/images/svg/index.js [hunk 5]
```diff
@@ -1,4 +1,5 @@
 export { default as facebook } from './facebook.svg';
+export { default as google } from './google.svg';
 export { default as logo } from './logo.svg';
 export { default as twitter } from './twitter.svg';
 export { default as logo2 } from './Logo2.svg';
```
Comment (p-funky) [inline src/assets/images/svg/index.js]: Let's rename this file to use lowercase please i.e. `logo2.svg`

File: src/styles/components/_signupForm.scss [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+.mx-5 {
+  margin-right: 3rem !important;
+  height: 4rem;
+}
+
+.fMYNqu,
+.bDmdDe,
+.eRNVUc,
+.ejpHkx,
+.hSTChk {
```
Comment (p-funky) [inline src/styles/components/_signupForm.scss]: What's up with these weird class names?

File: src/styles/styledComponents/tests/input.test.js [hunk 7]
```diff
@@ -4,17 +4,17 @@ import renderer from 'react-test-renderer';
 // eslint-disable-next-line import/no-unresolved
 import InputField from '../../../components/common/input/InputComponent';
 // eslint-disable-next-line import/no-unresolved
```
Comment (p-funky) [inline src/styles/styledComponents/tests/input.test.js]: We should get rid of these disables as well.

File: src/styles/styledComponents/tests/__snapshots__/input.test.js.snap [hunk 8]
```diff
@@ -2,6 +2,6 @@
 
 exports[`Custom Input styled-component renders correctly 1`] = `
 <input
-  className="sc-bdVaJa iyVdpk"
+  className="sc-bdVaJa SMGVi"
```
Comment (p-funky) [inline src/styles/styledComponents/tests/__snapshots__/input.test.js.snap]: If this class name is going to keep changing with every test run then it's definitely not scalable and not implemented right.

File: src/utils/index.js [hunk 9]
```diff
@@ -11,6 +9,15 @@ export const apiInstance = axios.create({
   }
 });
 
+export const validateToken = (token) => {
+  let isValid = false;
+  if (token) {
+    const decoded = jwtDecode(token);
+    isValid = decoded.exp > Date.now() / 1000 && decoded.exp && decoded;
```
Comment (p-funky) [inline src/utils/index.js]: You have this hierarchy wrong. Should be more like:
```suggestion
    isValid = decoded && decoded.exp && decoded.exp > Date.now() / 1000;
```
You should first check for the object before you check for a property within the object

File: src/utils/index.js [hunk 10]
```diff
@@ -38,20 +45,22 @@ export const renderPageLinks = (currentPage, numberOfPages) => {
...
+  const urlSearchParams = new URLSearchParams(window.location.search);
+  let page = parseInt(urlSearchParams.get('page'), 10);
+  if (Number.isNaN(page)) page = 1;
   component.setState({ currentPage: page });
 };
 
 export const filterArticlesByLikes = (articles) => {
   const topArticles = articles
-    .sort((a, b) => (a.likes > b.likes) ? 1 : ((b.likes < a.likes) ? -1 : 0)).slice(0, 7);
+    .sort((a, b) => (a.likes > b.likes ? 1 : b.likes < a.likes ? -1 : 0))
```
Comment (p-funky) [inline src/utils/index.js]: Let's at least correct the logic here before we create another function to refactor it. This should be:
```suggestion
    .sort((a, b) => (a.likes > b.likes ? 1 : a.likes < b.likes ? -1 : 0))
```

File: src/utils/validationSchemas.js [hunk 11]
```diff
@@ -0,0 +1,13 @@
+export const signUpFormSchema = {
+  email: /[-.\w]+@([\w-]+\.)+[\w-]{2,20}/,
+  username: /^(?=.*[a-zA-Z])\S{3,}$/,
```
Comment (p-funky) [inline src/utils/validationSchemas.js]: Your regex for username seems to be able to accept anything as long as a letter is present in it e.g. `_h*@'~`

File: src/utils/validationSchemas.js [hunk 12]
```diff
@@ -0,0 +1,13 @@
+export const signUpFormSchema = {
+  email: /[-.\w]+@([\w-]+\.)+[\w-]{2,20}/,
```
Comment (p-funky) [inline src/utils/validationSchemas.js]: Your regex for email does not seem to be as strict as you intend e.g`-@d1.cozsesfrw` email passes

File: src/utils/validationSchemas.js [hunk 13]
```diff
@@ -0,0 +1,13 @@
...
+  username: /^(?=.*[a-zA-Z])\S{3,}$/,
+  password: /^(?=.*[0-9])(?=.*[a-zA-Z])([a-zA-Z0-9]{8,})$/
```
Comment (p-funky) [inline src/utils/validationSchemas.js:4]: This expression does not cater for special characters like `-`, `_`, etc. Do you intend for the password to be strictly made up of alphabets and letters?

**Comment by p-funky** [comment]
Overall, checking out to this branch and running `npm run start-dev` I can't seem to access the project on the browser. Something seems to be failing in the build process.

File: src/App.js (hunk 4 above)
Comment (ozone4real) [reply src/App.js]: Okay. Would remove this.

File: src/assets/images/svg/index.js (hunk 5 above)
Comment (ozone4real) [reply src/assets/images/svg/index.js]: Okay

File: src/styles/components/_signupForm.scss (hunk 6 above)
Comment (ozone4real) [reply src/styles/components/_signupForm.scss]: I honestly don't know where these classNames came from.

File: src/utils/index.js (hunk 10 above)
Comment (ozone4real) [reply src/utils/index.js]: I didn't work on this. This is Tiku's work. The parenthesis were probably removed by my code formatter on save (I worked on some other functions in this file, but not this).

File: src/utils/validationSchemas.js (hunk 13 above)
Comment (ozone4real) [reply src/utils/validationSchemas.js:4]: The backend validation for password only permits alphabets and numbers. I didn't want to permit something on the front-end when it is not permitted on the backend.

File: src/utils/validationSchemas.js (hunk 11 above)
Comment (ozone4real) [reply src/utils/validationSchemas.js]: Before writing the validation for front-end, I looked at what the backend accepts and rejects and wrote the validations based on that. The backend would accept this. Should I make the validation more strict on the front-end?

File: src/utils/validationSchemas.js (hunk 12 above)
Comment (ozone4real) [reply src/utils/validationSchemas.js]: Yeah It seems. But I don't think we should be too strict with this.  Can't account for what every domain name ends with (whether .org, .com, .ng, .business, .investments ). But we know every domain name ends with a word. (Besides, we're doing email verification).

File: src/utils/index.js (hunk 9 above)
Comment (ozone4real) [reply src/utils/index.js]: True. Would change this.

File: __test__/actions/authActions.test.js (hunk 1 above)
Comment (ozone4real) [reply __test__/actions/authActions.test.js]: Thanks. I've corrected it

File: src/utils/index.js (hunk 10 above)
Comment (p-funky) [reply src/utils/index.js]: Errrmmm... if something changed within your PR who is responsible for the change but you? 🤷🏾‍♂️
Meanwhile, you missed the actual point. Take a closer look at the changes suggested in my comment.

File: src/utils/validationSchemas.js (hunk 11 above)
Comment (p-funky) [reply src/utils/validationSchemas.js]: I think you miss the point. Is it about what the back-end is currently accepting or about what it should actually be accepting?

File: src/utils/validationSchemas.js (hunk 12 above)
Comment (p-funky) [reply src/utils/validationSchemas.js]: `(Besides, we're doing email verification).` By this you mean on the backend?

Comment (p-funky) [reply src/utils/validationSchemas.js]: But the main point in my comment was that something like `-@d1.co` will pass for an email but what you have before the `@` is simply a symbol `-`. Do we want to allow such emails?

File: src/utils/validationSchemas.js (hunk 13 above)
Comment (p-funky) [reply src/utils/validationSchemas.js:4]: So we are cool with having our passwords to be strictly alphanumeric is what I was trying to verify.

File: src/utils/index.js (hunk 10 above)
Comment (ozone4real) [reply src/utils/index.js]: Okay

File: src/utils/validationSchemas.js (hunk 12 above)
Comment (ozone4real) [reply src/utils/validationSchemas.js]: Ohkay. I get now. I would work on this.

**Review by EfeAgare — DISMISSED** [review]

**Review by bobbylemm — DISMISSED** [review]

**Review by biolabalo — DISMISSED** [review]
LGTM

**Review by omenkish — DISMISSED** [review]

**Review by primuse — DISMISSED** [review]

File: __test__/components/SignupFormContainer.test.js [hunk 14]
```diff
@@ -60,13 +60,13 @@ describe('Sign up form container test', () => {
     wrapper.setState({
       errors: {
         username:
-          'Username must be at least 3 characters long and must include at least an alphabet'
+          'Username must start and end with an alphabet and must be 3 characters long. Special characters are not allowed'
```
Comment (p-funky) [inline __test__/components/SignupFormContainer.test.js]: Must be at least 3 characters long.

**Review by biolabalo — DISMISSED** [review]
LGTM

**Review by omenkish — DISMISSED** [review]

**Review by omenkish — APPROVED** [review]

**Review by EfeAgare — APPROVED** [review]

**Review by bobbylemm — APPROVED** [review]

**Pull request merged**